File: screen-reader/android/src/main/java/com/capacitorjs/plugins/screenreader/ScreenReaderPlugin.java
package com.capacitorjs.plugins.screenreader;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "ScreenReader")
public class ScreenReaderPlugin extends Plugin {

    public static final String EVENT_STATE_CHANGE = "stateChange";
    private ScreenReader screenReader;

    @Override
    public void load() {
        screenReader = new ScreenReader(getContext());
        screenReader.addStateChangeListener((enabled) -> {
            JSObject ret = new JSObject();
            ret.put("value", enabled);
            notifyListeners(EVENT_STATE_CHANGE, ret);
        });
    }

    @Override
    protected void handleOnDestroy() {
        screenReader.removeAllListeners();
    }

    @SuppressWarnings("unused")
    @PluginMethod
    public void isEnabled(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("value", screenReader.isEnabled());
        call.resolve(ret);
    }

    @SuppressWarnings("unused")
    @PluginMethod
    public void speak(PluginCall call) {
        String value = call.getString("value");
        String language = call.getString("language", "en");
        screenReader.speak(value, language);
        call.resolve();
    }
}

File: screen-reader/ios/Sources/ScreenReaderPlugin/ScreenReaderPlugin.swift
import Foundation
import Capacitor

@objc(ScreenReaderPlugin)
public class ScreenReaderPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ScreenReaderPlugin"
    public let jsName = "ScreenReader"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "speak", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isEnabled", returnType: CAPPluginReturnPromise)
    ]
    static let stateChangeEvent = "stateChange"

    override public func load() {
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.onVoiceOverStateChanged(notification:)),
                                               name: UIAccessibility.voiceOverStatusDidChangeNotification,
                                               object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    @objc func isEnabled(_ call: CAPPluginCall) {
        let enabled = UIAccessibility.isVoiceOverRunning

        call.resolve([
            "value": enabled
        ])
    }

    @objc func speak(_ call: CAPPluginCall) {
        guard let value = call.getString("value") else {
            call.reject("No value provided")
            return
        }

        if UIAccessibility.isVoiceOverRunning {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                UIAccessibility.post(notification: .announcement, argument: value)
            }
        }

        call.resolve()
    }

    @objc private func onVoiceOverStateChanged(notification: NSNotification) {
        notifyListeners(ScreenReaderPlugin.stateChangeEvent, data: [
            "value": UIAccessibility.isVoiceOverRunning
        ])
    }
}
